src/lib.rs

```diff
@@ -227,4 +323,109 @@ impl<T: Config> ManageRoles<T::AccountId> for Pallet<T> {
+        }
+        Self::burn(&user, amount)?;
+        Self::deposit_event(Event::<T>::Withdraw {
+            user: user.clone(),
+            amount,
+        });
+        Ok(())
+    }
+
+    fn transfer(from: &T::AccountId, to: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        if amount < T::MinimumAmount::get(){
+            return Err(Error::<T>::AmountTooSmall.into());
+        }
+        Accounts::<T>::mutate(&from, |balance| -> DispatchResult {
+            if balance.free >= amount {
+                balance.free -= amount;
+                Ok(())
+            } else {
+                Err(Error::<T>::InsufficientBalance.into())
+            }
+        })?;
+        Accounts::<T>::mutate(&to, |balance| {
+            balance.free = balance.free.saturating_add(amount);
+        });
+        Self::deposit_event(Event::Transfer {
+            from: from.clone(),
+            to: to.clone(),
+            amount,
+        });
+        Ok(())
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Burn some fund from a user's account.
+    fn burn(user: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        Self::ensure_role(&user, Role::Customer)?;
+        Accounts::<T>::mutate(&user, |balance| -> DispatchResult {
+            if balance.free >= amount {
+                balance.free -= amount;
+                Ok(())
+            } else {
+                Err(Error::<T>::InsufficientBalance.into())
+            }
+        })?;
+        TotalIssuance::<T>::mutate(|total| {
+            *total = total.saturating_sub(amount);
+        });
+        Ok(())
+    }
+
+    /// Mint some fund into a user's account.
+    fn mint(user: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        Self::ensure_role(&user, Role::Customer)?;
+
+        Accounts::<T>::mutate(&user, |balance| {
+            balance.free = balance.free.saturating_add(amount);
+        });
+        TotalIssuance::<T>::mutate(|total| {
+            *total = total.saturating_add(amount);
+        });
+        Ok(())
+    }
+
+    /// Integrity check: Ensure that the sum of all funds in balances matches total_issuance.
+    pub(crate) fn check_total_issuance() -> bool {
+        TotalIssuance::<T>::get()
+            == Accounts::<T>::iter()
+                .map(|(_, account)| account.free + account.reserved)
+                .sum()
+    }
+
+    /// Reaps funds from accounts that have balances below the Existential Deposit (ED).
+    /// Reaped funds are transferred to the Treasury account.
+    fn reap_accounts() {
+        let treasury = T::TreasuryAccount::get();
+        let ed = T::ExistentialDeposit::get();
+
+        <Accounts<T>>::iter().for_each(|(account_id, account)| {
+            if account.free < ed {
+                let reaped_amount = account.free;
+                Accounts::<T>::mutate(&treasury, |treasury_account| {
+                    treasury_account.free = treasury_account.free.saturating_add(reaped_amount);
```

Comment:
Try to do this with a single db write to the treasury's account
Also you should check for "total balance", free + reserved instead of just free, otherwise people with large amount of reserved fund is lost.
Should also transfer the reserved fund into the treasury too.

Reply:
try to use iterator functions:
```
Accounts::<T>::iter()
	.filter(|(_id, balance)| {
		balance.total() < T::ExistentialDeposit::get()
	})
	.for_each(|(id, balance)| {
		// transfer balance to treasury
                 // remove from storagemap 
	});

```